fsspec/implementations/dirfs.py

```diff
@@ -0,0 +1,311 @@
+from ..asyn import AsyncFileSystem
+
+
+class DirFileSystem(AsyncFileSystem):
+    def __init__(self, path, fs, *args, **storage_options):
+        """
+        Parameters
+        ----------
+        path: str
+            Path to the directory.
+        fs: AbstractFileSystem
+            An instantiated filesystem to wrap.
+        """
+        super().__init__(*args, **storage_options)
+
+        if self.asynchronous and not fs.async_impl:
+            raise ValueError("can't use asynchronous with non-async fs")
+
+        if self.asynchronous != fs.asynchronous:
+            raise ValueError("both dirfs and fs should be in the same sync/async mode")
```

Comment:
Decided not to try to wrap asynchronous fs with sync like reffs does, because it makes little sense in dirfs use case, as there is just one fs that we are dealing with and user expects to use dirfs in the same mode as fs it is wrapping. Though that also means that we could just mimic `fs.asynchronous` instead of passing it explicitly, but I kept it explicit for now.

Also, it is clear that some of these attributes (e.g. asynchronous/batch_size/etc) might be expected to be mirrorred by the dirfs, but they are obviously not right now. We can add them when someone needs them.

Reply:
Agreed to keep it simple until we see this being used and issues naturally arise. Since we are not aiming for generalised "mount" here, I suspect issues will be few, and this will mainly be used locally.